main.cc

```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
+#include <cassert>
+
+using namespace std;
+
+class String {
+public:
+    
+    String() :
+        String("", 0)
+    { }
+
+    String(const char* data) :
+        String(data, strlen(data))
+    { }
+
+    String(const String& s) :
+        String(s.data(), s.length())
+    { }
+    
+    String(const char* data, size_t length) {
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
+        m_data = new char[capacity];
```

Comment:
Consider *initializing* the members, rather than assigning them.

Reply:
Ok, will do.